### ISSUE TITLE
Write unit tests

### ISSUE THREAD

**Issue by jfoug** [magnumripper/JohnTheRipper#3576]
Going to write unit test 'program'.  This will start with unit testing all public functions in misc.c   I may extend to also test all public functions in common.c and there may be 'other' ancillory functions which should also unit test.  The program will be stand alone, BUT will be incorporated into makefile.in and makefile.legacy.   I will put the source into './src/unused', since it really is unused code (not part of JtR).  The tests will also be added to the CI, and likely be a simple build/run test using ASAN, so that buffer over/underflows will be caught.

I am pretty well along on the misc.c unit tests. I will get the unit test project into a PR soon, along with an updated misc.c with numerous issues corrected.

This will be in jumbo-2, but we can easily continue adding more functions into the unit-tests even after J2 is released.

**Comment by magnumripper** [comment]
> I will put the source into './src/unused', since it really is unused code (not part of JtR). 

We shouldn't mis-use "unused" like that. Perhaps it should be in the main tree (or better a new directory `src/test`) and be part of a normal build. If it fails on some obscure arch, build will bail out with a message (that should point user to GitHub as well as john-users ML) and we'll hopefully get to know it.

**Comment by jfoug** [comment]
I can easily move it to 'test'.  Right now, it only builds if you do a :

make misc_c_unit_test    or make -f Makefile.legacy/misc_c_unit_test

Now, I can easily rename to whatever.  NOTE, if is NOT part of a normal make, it is stand alone. It does use misc.c (and possibly memory.c), but it does not compile them to .o since it sets some #define constants prior to the compile that would fuxor up any .o files produced. It simply creates a stand alone exe, when when run, dumps out results (and lists errors as it sees them), and then returns 0 or 1 (depends on if any errors).  It would fit very well into a CI build, and should not take more than a trivial amount of time to run.  But once a unit test is in, then most changes to source in john which would cause any of the tested functions to fail, would also fail.  The only thing not tested, is things like order of initialization of external helper items (like the atoi[]/itoa[] stuff from common).  But the actual underlying workings of the unit tested functions will fail if we modify something (say making adjustments to headers, or some other external data item), or to the functions themselves.   Having the CI fail is a good thing.  It will stop some latent bug which may be hard to track down later, OR could cause a lot of red-herring debugging on things which really do not contain a bug.

Here is an example output:
```
Install missing libraries to get any needed features that were omitted.

Configure finished.  Now "make -s clean && make -sj4" to compile.

$ make misc_c_unit_test
gcc -o ../run/misc_c_unit_test -Wl,--no-undefined -O -g -fsanitize=address -fno-omit-frame-pointer -lasan -D_MISC_C_UNIT_TEST -D_JOHN_MISC_NO_LOG unused/misc_c_unit_test.c misc.c

$ ../run/misc_c_unit_test 
Performed testing on 22 functions
  test 0 test_real_error        -  no unit test yet
  test 1 test_real_error_msg    -  no unit test yet
  test 2 test_real_pexit        -  no unit test yet
  test 3 test_write_loop        -  no unit test yet
  test 4 _nontest_gen_fgetl_files-  Performed 3 tests 0.26 seconds used
  test 5 test_fgetl             -  Performed 18000 tests WITH 1 FAILURES!!! 3.49 seconds used
  test 6 test_fgetll            -  Performed 18000 tests 0.94 seconds used
  test 7 test_strncpy_pad       -  Performed 27 tests 0.01 seconds used
  test 8 test_strnfcpy          -  Performed 27 tests
  test 9 test_strnzcpy          -  Performed 27 tests
  test 10 test_strnzcpylwr      -  Performed 27 tests
  test 11 test_strnzcpyn        -  Performed 27 tests
  test 12 test_strnzcpylwrn     -  Performed 27 tests
  test 13 test_strnzcat         -  no unit test yet
  test 14 test_strnzcatn        -  Performed 2745 tests 0.01 seconds used
  test 15 test_strtokm          -  no unit test yet
  test 16 test_atou             -  Performed 200010 tests 0.46 seconds used
  test 17 test_jtr_itoa         -  Performed 8400420 tests 3.88 seconds used
  test 18 test_jtr_utoa         -  Performed 8400420 tests 4.41 seconds used
  test 19 test_jtr_lltoa        -  Performed 8400420 tests 5.36 seconds used
  test 20 test_jtr_ulltoa       -  Performed 8400420 tests 5.78 seconds used
  test 21 test_human_prefix     -  no unit test yet
```

Note, the misc.c is prior to the NULL byte 'fix' in fgetl (I think). I have not updated my VM Fedora, which is where I run any ASAN builds.

I also had a lot of ASAN leaks with the recently added fgetll. These are expected. I just cut/paste edited the code from fgetl test to fgetll test. I need to track if the pointer is different, and free it properly.  Then those memory leaks will go away.  Those are not leaks in the misc.c, but leaks in my unit test framework code.

**Comment by jfoug** [comment]
Ok, now that I have the bugs fixed in my unit tester for fgetl and fgetll (it was only marking up a single error, for the entire run of fgetl, and none for fgetll, since the error clear flag was not being reset for each line).

After getting it 'fixed', here are the tests:

```
before the fgetl/fgetll NULL byte code changes:

  test 4 _nontest_gen_fgetl_files -  Performed 3 tests 0.27 seconds used
  test 5 test_fgetl               -  Performed 18003 tests WITH 1800 FAILURES!!! 3.12 seconds used
  test 6 test_fgetll              -  Performed 18003 tests WITH 909 FAILURES!!! 0.94 seconds used

after the latest misc.c changes to fgetl and fgetll

  test 4 _nontest_gen_fgetl_files -  Performed 3 tests 0.31 seconds used
  test 5 test_fgetl               -  Performed 18003 tests 2.98 seconds used
  test 6 test_fgetll              -  Performed 18003 tests 0.84 seconds used
```

So I am much happier now with the code changes for the fgetl/fgetll.   The unit test code is MUCH more comprehensive than the prior small test files I had generated.  These tests tested long, short, over long (fgetl truncated and fgetll returned).  It validates that all data is properly handled no matter which function is called (within the specification of the function).

**Comment by magnumripper** [comment]
#3580 should have pointed to this issue. Merged now and seems to work like a champ.

The dependency on `default` is a bug, it's not needed and just makes a test-only run a lot slower:
```diff
diff --git a/src/Makefile.in b/src/Makefile.in
index fcb99e938..af3436222 100644
--- a/src/Makefile.in
+++ b/src/Makefile.in
@@ -797,7 +798,7 @@ tests/common.o:     common.c arch.h common.h memory.h os.h os-autoconf.h autoconfig.
 tests/memory.o:         memory.c arch.h misc.h jumbo.h autoconfig.h memory.h common.h johnswap.h os.h os-autoconf.h
        $(CC) -o tests/memory.o $(CFLAGS) -D_MISC_C_UNIT_TEST -D_JOHN_MISC_NO_LOG  memory.c
 
-unit_tests: $(UNIT_TEST_OBJS) default
+unit_tests: $(UNIT_TEST_OBJS)
        $(CC) -o ../run/unit_tests $(LDFLAGS) -D_MISC_C_UNIT_TEST -D_JOHN_MISC_NO_LOG $(UNIT_TEST_OBJS)
        @ echo "Now Running the Unit Tests"
        @ ../run/unit_tests
```

And with the above fixed, we can optionally do this:
```diff
@@ -185,6 +185,7 @@ endif
 #
 #########################################################
 default:
+       unit_tests
        $(MAKE) find_version
        @$(MAKE) $(PROJ) \
                JOHN_OBJS="$(JOHN_OBJS) @CC_ASM_OBJS@" \
```
The latter only adds ~10 seconds to total build time, but we always get proper unit testing when building

On a side note I'd personally much prefer the build target name as `unit-test` and not `unit_test`.

**Comment by jfoug** [comment]
> The dependency on default is a bug, it's not needed and just makes a test-only run a lot slower:

The dependency settings I originally had, were due to some commented out run lines AFTER the run of unit-tests.exe   Those tests required john.   It was still WIP, and since they are currently commented out, you are correct, the dependancy from unit-test back to default is NOT needed.

**Comment by jfoug** [comment]
NOTE,  I think the real goal (and real soon), will be to make unit-tests simply be a symlink to ../run/john

If I start testing our built in 'hash' code, then there simply will be TOO much john code to build.  Right now, it is pretty small, only a couple .c files, built to .o files out of the 'normal' location for them  BUT it it becomes 50 files (or 100, or more!!), then we certainly do not want this to be a stand alone project.

Note, if it is a symlink to john, then we can test ANY function  (even a single 'special' function in some format.  As long as the function is not static, and is part of john, then the unit-tests could handle it (if someone writes a test harness for it).    Also, as a symlink, there is only a second or less of extra time, ever, over a normal build.

**Comment by magnumripper** [comment]
Sounds good but I have yet to look into why you are building "special" versions of some files, e.g. with `-D_MISC_C_UNIT_TEST`... I presume that is a showstopper.

**Comment by jfoug** [comment]
It 'wa

> building "special" versions of some files, e.g. with `-D_MISC_C_UNIT_TEST`... I presume that is a showstopper.

It was, at one time. But now I have put enough code into the project, that this define is actually not even in code any more:

```
$  grep _MISC_C_UNIT_TEST *.[ch]
```

I simply have not removed it.   It was there for things like the VC snprintf mess.  For VC always using _aligned_alloc, etc..  But since memory is now part of the build, both of these issues are handled, and this define should simply go away.

Now if (or when) it simply becomes a john symlink, then there are no special defines at all needed. It is just part of john, and will compile simply as another module linked to the core john.

**Comment by jfoug** [comment]
Completed with #3586 and #3580 

 NOTE, there will always be additional things which CAN be added to be unit tested.  These enhancements to the unit-tests code can be their own issue/PR.    The basis for this issue is now done.

**Comment by magnumripper** [comment]
Good point. I do think we should run them, but they could be almost quiet as long as nothing fails.

**Comment by jfoug** [comment]
@solardiz   I initially started out with that approach (unix mentality : if no errors stfu).   However, as more and more tests are added, do we want to sit and see 'nothing' for minutes?   Right now, it is pretty quick (under a minute, and usually under 30s).  I can easily detect and pass through the -s switch to make (I think), and perform tests in a quiet manner.   I guess as long as the end user that this step "can" take a while, and to not be alarmed, then all is OK.

**Comment by solardiz** [comment]
If it starts taking annoyingly long for the tests to complete, that's a reason not to run them by default, or maybe to run a light version by default (e.g., only a few tests per function, not thousands). Maybe we're already at a point where we should do that.

**Comment by magnumripper** [comment]
Current tests take over 7 seconds. Given what we are testing that's quite some time already, we should try to make it faster and it should definitely never go beyond 10 seconds IMHO. On a side note I can't see why we'd test SHA-2 at all - that's included in the format self-tests. The unit tests should be for things not tested otherwise, like the misc.c and common.c stuff we do test (and jumbo.c to be added).

I think we should print something like `Performing unit tests...` before running them, and as long as nothing failed we add a `DONE in 7.29 s` on same line.

**Comment by magnumripper** [comment]
> Current tests take over 7 seconds

From the look of it this might be including *building* the tests though, right? Are we not using the actual misc.o and common.o of the build? If we are, that time can be subtracted.

**Comment by jfoug** [comment]
Ok,

```
$ make
MAKEFLAGS=
$ make -s
MAKEFLAGS=s
$ make -sj4
MAKEFLAGS=s -j4 --jobserver-auth=3,4
```
The MAKEFLAGS is env var.  So it's trivial to simply walk the ```main(,,const char **envp)``` array, capture a MAKEFLAGS value, parse it for a stand alone 's', and if seen, we know to behave in the STFU (almost) manner which the -s wants.   Easy, peasy.

Also this would be a 'switch' to tell us to run a light version (if we deem that what we want).

Most functions take no extra (observed) time based on number of functions. However, some do have many functions run (and really 'need' it that way).  Initially, I had things like itoa testing, testing -8000000 to 7ffffffff  (i.e. entire range).  I knew I was not going to put it out that way, but wanted to FULL test, as I was developing. NOW, that code tests edge conditions, and makes sure that worse case length jumping is done i.e. testing from 1 to 100010 gets 6 digit length increases. Then test a bit before and after 1 billion.  That gets the most bang for the buck, without having to test 66677789 knowing that 66677788 (and all prior) also worked.

**Comment by solardiz** [comment]
We can have extensive unit tests as part of `make check`, like other projects have. It's weird to have default-enabled unit tests outside of that make target.

**Comment by magnumripper** [comment]
Never mind the `-s` flag, we can do what I said regardless.

**Comment by solardiz** [comment]
The more I think of it, the more I dislike the idea of running any unit tests along with normal builds. This goes against packagers' reasonable expectations, where `make check` is supposed to be a separate target performing work that doesn't overlap with what's done on a simple `make`.

**Comment by jfoug** [comment]
@solardiz  I initially built it this way, but it was changed by @magnumripper   I also balked a bit, but once I realized that the make was simply postponed 'the first time' to allow unit tests to pass, I changed my thought pattern.

We do not have a make install.   Normally it is at the make install target, which the install will fail, if the unit-testing fails, and unit testing is force to be done.  Most makes can 'force' install even if testing fails. 

Now, if you would like to revert this, then I can make it so, add the check targets, and change the CI scripts to run ```make check``` 'first'.

**Comment by solardiz** [comment]
Adding a `make install` is also something we should revisit, separately.

I disagree that "normally" testing is invoked by `make install`, and that install has to be forced if tests fail. This is the case with CPAN, but I don't recall any other example (I'm sure there are some, but I doubt this is prevalent). In packages we build for Owl, `make install` certainly works without a prior `make check` and doesn't invoke that target on its own. This is how we're able to do quicker without-testing (re)builds.

Yes, please introduce a separate `make check` and invoke it from CI. Thanks!

**Comment by claudioandre-br** [comment]
There is a `make check`. It does
* ../run/john --list=build-info
* ../run/john --test=0 --verbosity=2

Remember  that the test fails on OpenCL (AMD drivers). Some formats also fails on OpenCL CPU.

**Comment by jfoug** [comment]
Added a new issue for make-check.  If we need to discuss things further, which are NOT specifically part of the unit-tests project (hooking it into make really falls into #3614), then we should post on that issue instead.

Issues and requirements FOR unit-tests still should re-open this topic.  I will be leaving it open for now, since I will have code additions to make it 'honor' the ``` -s ``` flag of make.  But how to hook it into the actual make process probably should be done on the other issue topic.

**Comment by magnumripper** [comment]
Again, I strongly advice against caring the slightest about `-s`. It only adds unneeded complexity and it's not even logical at all IMO. KISS applies here.

Since we're now doing this as a separate target, I think we could be 100% silent as long as nothing fails, and very verbose about anything failing.

**Comment by solardiz** [comment]
@jfoug I agree with magnum - no need to treat `-s` special when we're moving the self-tests to `make check`. It is OK for output from `make check` to be verbose.
